Inductor cpp wrapper: fix optional tensor input

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106847

Fix cpp wrapper failure on `clip` in Torchbench:
```
RuntimeError: tensor does not have a device
```

An `optional<at::Tensor>` variable with value equal to `at::Tensor()` will be considered as _contains value_. When it's converted to `bool`, it returns `true`. While for `None` in python, when converting it to `bool`, `false` is returned.
Fix it to be an optional variable that _does not contain a value_.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

chunyuan-w committed:
Inductor cpp wrapper: fix optional tensor input

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix optional tensor input"


Fix cpp wrapper failure on `clip` in Torchbench:
```
RuntimeError: tensor does not have a device
```

`at::Tensor()` is not equivalent to `None` in python.
An `optional<at::Tensor>` variable with value equal to `at::Tensor()` will be considered as _contains value_. When it's converted to `bool`, it returns `true`. While for `None` in python, when converting it to `bool`, `false` is returned.
Fix it to be an optional variable that _does not contain a value_.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix optional tensor input"


Fix cpp wrapper failure on `clip` in Torchbench:
```
RuntimeError: tensor does not have a device
```

`at::Tensor()` is not equivalent to `None` in python.
An `optional<at::Tensor>` variable with value equal to `at::Tensor()` will be considered as _contains value_. When it's converted to `bool`, it returns `true`. While for `None` in python, when converting it to `bool`, `false` is returned.
Fix it to be an optional variable that _does not contain a value_.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix optional tensor input"


Fix cpp wrapper failure on `clip` in Torchbench:
```
RuntimeError: tensor does not have a device
```

`at::Tensor()` is not equivalent to `None` in python.
An `optional<at::Tensor>` variable with value equal to `at::Tensor()` will be considered as _contains value_. When it's converted to `bool`, it returns `true`. While for `None` in python, when converting it to `bool`, `false` is returned.
Fix it to be an optional variable that _does not contain a value_.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

chunyuan-w commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1294,7 +1299,7 @@ def val_to_str(self, val):
 
         if val is None:
-            return self.none_str
+            return self.optional_tensor_str
```
We'll need both of them.
When `None` is used as a `Constant` buffer, we do need to generate `at::Tensor()` for the cpp wrapper.
https://github.com/pytorch/pytorch/blob/e4e9aa28a7c06819b233753a93eff2d65679c627/torch/_inductor/ir.py#L2171-L2173

For example for the below backward graph, the code using python wrapper is as below where the return of the `call` function is `return (None, None, None, buf0, )`.
```log
def call(args):
    primals_1, primals_2, primals_3, sym_size, tangents_1 = args
    args.clear()
    s0 = primals_1
    s1 = primals_2
    s2 = primals_3
    assert_size_stride(tangents_1, (1, s0, (-1) + s1, s2), (((-1)*s0*s2) + (s0*s1*s2), ((-1)*s2) + (s1*s2), s2, 1))
    buf0 = empty_strided((1, s0, s1, s2), (s0*s1*s2, s1*s2, s2, 1), device='cpu', dtype=torch.float32)
    cpp_fused_slice_backward_0(c_void_p(tangents_1.data_ptr()), c_void_p(buf0.data_ptr()), c_long(s0), c_long(s1), c_long(s2))
    del tangents_1
    return (None, None, None, buf0, )


def benchmark_compiled_module(times=10, repeat=10):
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    primals_1 = 2
    primals_2 = 16
    primals_3 = 4
    sym_size = 15
    tangents_1 = rand_strided((1, 2, 15, 4), (120, 60, 4, 1), device='cpu', dtype=torch.float32)
    return print_performance(lambda: call([primals_1, primals_2, primals_3, sym_size, tangents_1]), times=times, repeat=repeat)
```


The corresponding cpp wrapper code will be as below with the return of `std::vector<at::Tensor>` type to be `return {at::Tensor(), at::Tensor(), at::Tensor(), buf0};`:
```log
std::vector<at::Tensor> inductor_entry_cpp(const std::vector<at::Tensor>& args) {
    long primals_1 = args[0].item<long>();
    long primals_2 = args[1].item<long>();
    long primals_3 = args[2].item<long>();
    long sym_size = args[3].item<long>();
    at::Tensor tangents_1 = args[4];
    auto s0 = primals_1;
    auto s1 = primals_2;
    auto s2 = primals_3;

    c10::optional<at::Scalar> optional_scalar;
    c10::optional<c10::string_view> optional_string;
    c10::optional<at::Layout> optional_layout;
    c10::optional<at::Tensor> optional_tensor;
    torch::List<c10::optional<at::Scalar>> optional_list;
        auto buf0 = at::empty_strided({1L, s0, s1, s2}, {s0*s1*s2, s1*s2, s2, 1L}, at::TensorOptions(at::kCPU).dtype(at::kFloat));
        cpp_fused_slice_backward_0((float*)(tangents_1.data_ptr()), (float*)(buf0.data_ptr()), s0, s1, s2);
        tangents_1.reset();
        return {at::Tensor(), at::Tensor(), at::Tensor(), buf0};
}
'''
)

module = CppWrapperCodeCache.load(cpp_wrapper_src, 'inductor_entry_cpp', 'choudiaq2rjxxab2tzkhrneb5msxtjk37p64b4falmwuhyrbzn2a', False)

def _wrap_func(f):
    def g(args):
        args_tensor = [arg if isinstance(arg, torch.Tensor) else torch.tensor(arg) for arg in args]
        return f(args_tensor)
    return g
call = _wrap_func(module.inductor_entry_cpp)


def benchmark_compiled_module(times=10, repeat=10):
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    primals_1 = 2
    primals_2 = 16
    primals_3 = 4
    sym_size = 15
    tangents_1 = rand_strided((1, 2, 15, 4), (120, 60, 4, 1), device='cpu', dtype=torch.float32)
    return print_performance(lambda: call([primals_1, primals_2, primals_3, sym_size, tangents_1]), times=times, repeat=repeat)
```

While here it is used in `codegen_args` to generate optional args, we'll need to generate an optional value.

jgong5 replied:
Thanks. If this change only applies to the args, can we change the method name from `val_to_str` to `val_to_arg_str`? Also, suggest to add a comment on why we return `optional_tensor_str` on `None` here.

chunyuan-w replied:
Function name has been changed and comment has been added.

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix optional tensor input"


Fix cpp wrapper failure on `clip` in Torchbench:
```
RuntimeError: tensor does not have a device
```

An `optional<at::Tensor>` variable with value equal to `at::Tensor()` will be considered as _contains value_. When it's converted to `bool`, it returns `true`. While for `None` in python, when converting it to `bool`, `false` is returned.
Fix it to be an optional variable that _does not contain a value_.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix optional tensor input"


Fix cpp wrapper failure on `clip` in Torchbench:
```
RuntimeError: tensor does not have a device
```

An `optional<at::Tensor>` variable with value equal to `at::Tensor()` will be considered as _contains value_. When it's converted to `bool`, it returns `true`. While for `None` in python, when converting it to `bool`, `false` is returned.
Fix it to be an optional variable that _does not contain a value_.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

chunyuan-w commented:
@pytorchbot rebase

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix optional tensor input"


Fix cpp wrapper failure on `clip` in Torchbench:
```
RuntimeError: tensor does not have a device
```

An `optional<at::Tensor>` variable with value equal to `at::Tensor()` will be considered as _contains value_. When it's converted to `bool`, it returns `true`. While for `None` in python, when converting it to `bool`, `false` is returned.
Fix it to be an optional variable that _does not contain a value_.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

chunyuan-w commented:
@pytorchbot merge

chunyuan-w commented:
@pytorchbot merge
